### ISSUE TITLE
why dapper return Int64 ?

### ISSUE THREAD

**Issue by cdmin207078** [StackExchange/dapper-dot-net#524]
```csharp
var sql = @"select count(1) from Products;
                            select * from Products order by ProductId asc limit @ItemsPerPage Offset @Offset";

                using (var multi = _db.Connection.QueryMultiple(sql, new { ItemsPerPage = page.ItemsPerPage, Offset = (page.CurrentPage - 1) * page.ItemsPerPage }))
                {
                    var totalitems = multi.Read<int>().Single();
                    var data = multi.Read<Product>().ToList();
                }
```
Above is wrong. and push a  `System.InvalidCastException'  

  If I change the following way ,it's correct ?
```csharp
var totalitems = multi.Read<Int64>().Single();
```
Is a Bug ?

However, the query does not have this problem.

**Comment by mgravell** [comment]
What version of dapper are you using? I believe this is already addressed
in current builds.

**Comment by cdmin207078** [comment]
Thank you for your reply.
I got it from nuget.

You can try the following case:
```csharp
var sql = "select 1 as A;select 2 as B";
 using (var multi = _db.Connection.QueryMultiple(sql))
 {
         var a = multi.Read<int>().Single();
         var b = multi.Read<int>().Single();
 }
```
![image](https://cloud.githubusercontent.com/assets/1762278/15602410/b0f6f066-2427-11e6-9a64-8c957f8fe0ee.png)

Is it because I am using the database is **SQLite**

**Comment by mgravell** [comment]
Again, which *version* of dapper are you using? It would be worth trying
the latest. Because of .net-core, many of the recent builds have been
marked pre-release, but they are absolutely fine for regular .net (it is
only the .net-core part that is really pre-release). I strongly suggest you
try on the latest version (1.50.0-rc2a), and see if it now works. I suspect
it will work fine.

To install a pre-release version from nuget, either use:

    Install-Package Dapper -Pre

(in the package manager console), or if you are using the IDE UI tools,
check the "Include prerelease" checkbox.


On 27 May 2016 at 09:28, 天雨流芳 <notifications@github.com> wrote:

> Thank you for your reply.
> I got it from nuget.
>
> You can try the following case:
>
> var sql = "select 1 as A;select 2 as B";
>  using (var multi = _db.Connection.QueryMultiple(sql))
>  {
>          var a = multi.Read<int>().Single();
>          var b = multi.Read<int>().Single();
>  }
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/1762278/15602410/b0f6f066-2427-11e6-9a64-8c957f8fe0ee.png>
>
> Is it because I am using the database is *SQLite*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/StackExchange/dapper-dot-net/issues/524#issuecomment-222090193>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AABDsFz48tsXJEmxZ33OQYQUCuAgjSXOks5qFqs5gaJpZM4IoQbV>
> .
>



-- 
Regards,

Marc

**Comment by mgravell** [comment]
Now that I'm back at a PC, I've tested this locally, and I can confirm
there is a bug here; the following is fine:

    i = db.Query<int>("select cast(1 as bigint);").Single();
    j = db.QuerySingle<int>("select cast(1 as bigint);");

However, the similar via the multi-reader is broken:

                using (var multi = db.QueryMultiple("select cast(1 as
bigint); select cast(2 as bigint)"))
                {
                    i = multi.Read<int>().Single(); // breaks
                    j = multi.ReadSingle<int>();
                }

I will investigate and fix.

On 27 May 2016 at 09:53, Marc Gravell <marc.gravell@gmail.com> wrote:

> Again, which *version* of dapper are you using? It would be worth trying
> the latest. Because of .net-core, many of the recent builds have been
> marked pre-release, but they are absolutely fine for regular .net (it is
> only the .net-core part that is really pre-release). I strongly suggest you
> try on the latest version (1.50.0-rc2a), and see if it now works. I suspect
> it will work fine.
>
> To install a pre-release version from nuget, either use:
>
>     Install-Package Dapper -Pre
>
> (in the package manager console), or if you are using the IDE UI tools,
> check the "Include prerelease" checkbox.
>
>
> On 27 May 2016 at 09:28, 天雨流芳 <notifications@github.com> wrote:
>
>> Thank you for your reply.
>> I got it from nuget.
>>
>> You can try the following case:
>>
>> var sql = "select 1 as A;select 2 as B";
>>  using (var multi = _db.Connection.QueryMultiple(sql))
>>  {
>>          var a = multi.Read<int>().Single();
>>          var b = multi.Read<int>().Single();
>>  }
>>
>> [image: image]
>> <https://cloud.githubusercontent.com/assets/1762278/15602410/b0f6f066-2427-11e6-9a64-8c957f8fe0ee.png>
>>
>> Is it because I am using the database is *SQLite*
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/StackExchange/dapper-dot-net/issues/524#issuecomment-222090193>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AABDsFz48tsXJEmxZ33OQYQUCuAgjSXOks5qFqs5gaJpZM4IoQbV>
>> .
>>
>
>
>
> --
> Regards,
>
> Marc
>



-- 
Regards,

Marc

**Comment by cdmin207078** [comment]
Well, thank you for your answer .

**Comment by mgravell** [comment]
I'm re-opening this to track as I fix it.

**Comment by Amateursss** [comment]
can you speak english...

**Comment by mgravell** [comment]
This will be fixed in any build **after** 1.50.0-rc2a. Until then: you may have to either a: cast the data in your RDBMS, or b: ask for a `long`, then cast to an `int` in C#.
